File: luna/src/nodes/controlNodes/range/range.cpp
#include "range.h"
#include <core/scripting/scriptingEngine.h>
#include <core/debug/typedefs.h>
#include <core/object/methodDB.h>
namespace luna 
{
	namespace nodes
	{
		

		static void RangeSetAllowGreater(entt::entity ObjectId, bool value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setAllowGreater(value);
		}

		
		static bool RangeIsGreaterAllowed(entt::entity ObjectId) 
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).isGreaterAllowed();
		}

		
		static void RangeSetAllowLesser(entt::entity ObjectId, bool value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setAllowLesser(value);
		}
		
		static bool RangeIsLesserAllowed(entt::entity ObjectId)
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).isLesserAllowed();
		}
		
		static void RangeSetExpRatio(entt::entity ObjectId, bool value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setExpRatio(value);
		}
		
		static bool RangeIsRatioExp(entt::entity ObjectId)
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).isRatioExp();
		}
		
		static void RangeSetMax(entt::entity ObjectId, float value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setAllowGreater(value);
		}
		
		static float RangeGetMax(entt::entity ObjectId) 
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).getMax();
		}
		
		static void RangeSetMin(entt::entity ObjectId, float value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setAllowGreater(value);
		}
		
		static float RangeGetMin(entt::entity ObjectId)
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).getMin();
		}
		
		static void RangeSetPage(entt::entity ObjectId, float value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setPage(value);

		}
		
		static float RangeGetPage(entt::entity ObjectId)
		{
			return 	range(ObjectId, scripting::scriptingEngine::getContext()).getPage();

		}
		
		static void RangeSetAsRatio(entt::entity ObjectId, float value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setAllowGreater(value);
		}
		
		static float RangeGetAsRatio(entt::entity ObjectId)
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).getAsRatio();
		}
		
		static void RangeSetUseRoundedValues(entt::entity ObjectId, bool value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setAllowGreater(value);

		}
		
		static bool RangeIsUsingRoundedvalues(entt::entity ObjectId)
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).isUsingRoundedvalues();

		}
		
		static void RangeSetStep(entt::entity ObjectId, float value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setStep(value);

		}
		
		static float RangeGetStep(entt::entity ObjectId)
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).getStep();
		}
		
		static void RangeSetValue(entt::entity ObjectId, float value) 
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setValue(value);

		}
		
		static float RangeGetValue(entt::entity ObjectId)
		{
			return range(ObjectId, scripting::scriptingEngine::getContext()).getValue();
		}
		
		static void RangeSetValueNoSignal(entt::entity ObjectId, float value)
		{
			range(ObjectId, scripting::scriptingEngine::getContext()).setValueNoSignal(value);
		}



		range::range(entt::entity handle, luna::scene* scene) : controlNode(handle,scene)
		{
		}
		void range::init(luna::scene* scene)
		{
			controlNode::init(scene);
			addComponent<rangeComponent>();
			LN_CLASS_TYPE_NAME(range);
		}
		void range::bindMethods()
		{
			LN_ADD_INTERNAL_CALL(range, RangeSetAllowGreater);
			LN_ADD_INTERNAL_CALL(range, RangeIsGreaterAllowed);

			LN_ADD_INTERNAL_CALL(range, RangeSetAllowLesser);
			LN_ADD_INTERNAL_CALL(range, RangeIsLesserAllowed);

			LN_ADD_INTERNAL_CALL(range, RangeSetExpRatio);
			LN_ADD_INTERNAL_CALL(range, RangeIsRatioExp);

			LN_ADD_INTERNAL_CALL(range, RangeSetMax);
			LN_ADD_INTERNAL_CALL(range, RangeGetMax);

			LN_ADD_INTERNAL_CALL(range, RangeGetMin);
			LN_ADD_INTERNAL_CALL(range, RangeSetMin);

			LN_ADD_INTERNAL_CALL(range, RangeSetPage);
			LN_ADD_INTERNAL_CALL(range, RangeGetPage);

			LN_ADD_INTERNAL_CALL(range, RangeGetAsRatio);
			LN_ADD_INTERNAL_CALL(range, RangeSetAsRatio);

			LN_ADD_INTERNAL_CALL(range, RangeIsUsingRoundedvalues);
			LN_ADD_INTERNAL_CALL(range, RangeSetUseRoundedValues);

			LN_ADD_INTERNAL_CALL(range, RangeGetStep);
			LN_ADD_INTERNAL_CALL(range, RangeSetStep);

			LN_ADD_INTERNAL_CALL(range, RangeGetValue);
			LN_ADD_INTERNAL_CALL(range, RangeSetValue);
			LN_ADD_INTERNAL_CALL(range, RangeSetValueNoSignal);
		}
		void range::setAllowGreater(bool value)
		{
			getComponent<rangeComponent>().allow_greater = value;
		}
		bool range::isGreaterAllowed()
		{
			return getComponent<rangeComponent>().allow_greater;
		}
		void range::setAllowLesser(bool value)
		{
			getComponent<rangeComponent>().allow_lesser = value;
		}
		bool range::isLesserAllowed()
		{
			return getComponent<rangeComponent>().allow_lesser;
		}
		void range::setExpRatio(bool value)
		{
			getComponent<rangeComponent>().exp_ratio = value;
		}
		bool range::isRatioExp()
		{
			return getComponent<rangeComponent>().exp_ratio;
		}
		void range::setMax(float value)
		{
			getComponent<rangeComponent>().max = value;
			LN_EMIT_SIGNAL("Changed");
		}
		float range::getMax()
		{
			return getComponent<rangeComponent>().max;
		}
		void range::setMin(float value)
		{
			if (getComponent<rangeComponent>().min == value) {
				return;
			}

			getComponent<rangeComponent>().min = value;
			getComponent<rangeComponent>().max = MAX(getComponent<rangeComponent>().max, getComponent<rangeComponent>().min);
			getComponent<rangeComponent>().page = CLAMP(getComponent<rangeComponent>().page, 0, getComponent<rangeComponent>().max - getComponent<rangeComponent>().max); 
			LN_EMIT_SIGNAL("Changed");
		}
		float range::getMin()
		{
			return getComponent<rangeComponent>().min;
		}
		void range::setPage(float value)
		{
			double pageValidated = CLAMP(value, 0, getMax() - getMin());
			if (getPage() == pageValidated) {
				return;
			}

			getComponent<rangeComponent>().page = pageValidated;
			setValue(getValue());
		
			LN_EMIT_SIGNAL("Changed");
		}
		float range::getPage()
		{
			return getComponent<rangeComponent>().page;
		}

		void range::setAsRatio(float value)
		{
			double v;

			if (isRatioExp() && getMin() >= 0) {
				double exp_min = getMin() == 0 ? 0.0 : glm::log(getMin()) / glm::log((double)2);
				double exp_max = glm::log(getMax()) / glm::log((double)2);
				v = glm::pow(2, exp_min + (exp_max - exp_min) * value);
			}
			else {
				double percent = (getMax() - getMin()) * value;
				if (getStep() > 0) {
					double steps = round(percent / getStep());
					v = steps * getStep() + getMin();
				}
				else {
					v = percent + getMin();
				}
			}
			v = CLAMP(v, getMin(), getMax());
			setValue(v);
		}

		float range::getAsRatio()
		{
			if (isEqualApprox(getMax(), getMin(),0.5f)) {
				// Avoid division by zero.
				return 1.0;
			}

			if (isRatioExp() && getMin() >= 0) {
				double exp_min = getMin() == 0 ? 0.0 : glm::log(getMin()) / glm::log((double)2);
				double exp_max = glm::log(getMax()) / glm::log((double)2);
				float value = CLAMP(getValue(), getMin(), getMax());
				double v = glm::log(value) / glm::log((double)2);

				return CLAMP((v - exp_min) / (exp_max - exp_min), 0, 1);
			}
			else {
				float value = CLAMP(getValue(), getMin(), getMax());
				return CLAMP((value - getMin()) / (getMax() - getMin()), 0, 1);
			}
		
		}

		void range::setUseRoundedValues(bool value)
		{
			getComponent<rangeComponent>().rounded = value;
		}

		bool range::isUsingRoundedvalues()
		{
			return getComponent<rangeComponent>().rounded;
		}

		void range::setStep(float value)
		{
			if (getStep() == value) return;
			getComponent<rangeComponent>().step = value;
			LN_EMIT_SIGNAL("Changed");
		}

		float range::getStep()
		{
			return getComponent<rangeComponent>().step;
		}

		void range::setValue(float value)
		{
			double prev_val = getValue();
			setValueNoSignal(value);

			if (getValue() != prev_val) {
				LN_EMIT_SIGNAL("ValueChanged", value);
			}
		}

		float range::getValue()
		{
			return getComponent<rangeComponent>().val;
		}

		void range::setValueNoSignal(float value)
		{
			if (!isfinite(value)) {
				return;
			}

			if (getStep() > 0) {
				value = glm::round((value - getMin()) / getStep()) * getStep() + getMin();
			}

			if (isUsingRoundedvalues()) {
				value = glm::round(value);
			}

			if (!isGreaterAllowed() && value > getMax() - getPage()) {
				value = getMax() - getPage();
			}

			if (!isLesserAllowed() && value < getMin()) {
				value = getMin();
			}

			if (getValue() == value) {
				return;
			}

			getComponent<rangeComponent>().val = value;

		}



	}
}

File: scriptCore/src/controlNodes/range/range.cs
using Luna;
using System;
using System.Runtime.CompilerServices;

namespace Luna
{
    public class Range : ControlNode
    {
        #region public methods
        protected void SetAllowGreater(bool value) { RangeSetAllowGreater(ObjectId, value); }
        protected bool IsGreaterAllowed() { return RangeIsGreaterAllowed(ObjectId); }

        protected void SetAllowLesser(bool value) { RangeSetAllowLesser(ObjectId, value); }
        protected bool IsLesserAllowed() { return RangeIsLesserAllowed(ObjectId); }

        protected void SetExpRatio(bool value) { RangeSetExpRatio(ObjectId, value); }
        protected bool GetExpRatio() { return RangeIsRatioExp(ObjectId); }

        protected void SetMaxValue(float value) { RangeSetMax(ObjectId, value); }
        protected float GetMaxValue() { return RangeGetMax(ObjectId); }

        protected void SetMinValue(float value) { RangeSetMin(ObjectId, value); }
        protected float GetMinValue() { return RangeGetMin(ObjectId); }

        protected void SetPage(float value) { RangeSetPage(ObjectId, value); }
        protected float GetPage() { return RangeGetPage(ObjectId); }


        #endregion

        #region Signals
        /// <summary>
        /// Emitted when MinValue, MaxValue, Page, or Step change.
        /// </summary>
        [Signal]
        public virtual void Changed() { }
        /// <summary>
        /// Emitted when value changes. When used on a Slider, this is called continuously while dragging (potentially every frame). If you are performing an expensive operation in a function connected to value_changed, consider using a debouncing Timer to call the function less often.
        /// </summary>
        /// <remarks>
        /// Unlike signals such as LineEdit.text_changed, value_changed is also emitted when value is set directly via code.
        /// </remarks>
        /// <param name="value"></param>
        [Signal]
        public virtual void ValuecChanged(float value) { }
        #endregion

        #region Dummy Fields
        /// <summary>
        /// If true, value may be greater than MaxValue.
        /// </summary>
        protected bool AllowGreater { get => RangeIsGreaterAllowed(ObjectId); set => RangeSetAllowGreater(ObjectId, value); }
        /// <summary>
        /// If true, value may be less than MinValue.
        /// </summary>
        protected bool AllowLesser { get => RangeIsLesserAllowed(ObjectId); set => RangeSetAllowLesser(ObjectId, value); }
        /// <summary>
        /// If true, and minValue is greater than 0, value will be represented exponentially rather than linearly.
        /// </summary>
        protected bool ExpEdit { get => RangeIsRatioExp(ObjectId); set => RangeSetExpRatio(ObjectId, value); }
        /// <summary>
        /// Maximum value. Range is clamped if value is greater than MaxValue.
        /// </summary>
        protected float MaxValue { get => RangeGetMax(ObjectId); set => RangeSetMax(ObjectId, value); }
        /// <summary>
        /// Minimum value. Range is clamped if value is less than MinVlue.
        /// </summary>
        protected float MinValue { get => RangeGetMin(ObjectId); set => RangeSetMin(ObjectId, value); }
        /// <summary>
        /// Page size. Used mainly for ScrollBar. ScrollBar's length is its size multiplied by page over the difference between MinValue and MaxValue.
        /// </summary>
        protected float Page { get => RangeGetPage(ObjectId); set => RangeSetPage(ObjectId, value); }
        /// <summary>
        /// The value mapped between 0 and 1.
        /// </summary>
        protected float Ratio { get => RangeGetAsRatio(ObjectId); set => RangeSetAsRatio(ObjectId, value); }
        /// <summary>
        /// If greater than 0, value will always be rounded to a multiple of this property's value. If rounded is also true, value will first be rounded to a multiple of this property's value, then rounded to the nearest integer.
        /// </summary>
        protected float Step { get => RangeGetStep(ObjectId); set => RangeSetStep(ObjectId, value); }
        /// <summary>
        /// Range's current value. Changing this property (even via code) will trigger <see cref="Changed"/> signal. Use SetValueNoSignal if you want to avoid it.
        /// </summary>
        protected float Value { get => RangeGetValue(ObjectId);set => RangeSetValue(ObjectId, value); }
        #endregion
        #region internal calls
        [MethodImplAttribute(MethodImplOptions.InternalCall)]

        static extern void RangeSetAllowGreater(ulong ObjectId, bool value);

        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern bool RangeIsGreaterAllowed(ulong ObjectId);

        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetAllowLesser(ulong ObjectId,bool value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern bool RangeIsLesserAllowed(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetExpRatio(ulong ObjectId,bool value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern bool RangeIsRatioExp(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetMax(ulong ObjectId,float value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern float RangeGetMax(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetMin(ulong ObjectId, float value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern float RangeGetMin(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetPage(ulong ObjectId, float value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern float RangeGetPage(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetAsRatio(ulong ObjectId, float value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern float RangeGetAsRatio(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetUseRoundedValues(ulong ObjectId, bool value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern bool RangeIsUsingRoundedvalues(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetStep(ulong ObjectId, float value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern float RangeGetStep(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetValue(ulong ObjectId, float value);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern float RangeGetValue(ulong ObjectId);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        static extern void RangeSetValueNoSignal(ulong ObjectId, float value);
        #endregion
        #region dummy fields

        #endregion
        protected Range() { }
        internal Range(ulong ObjectId) : base(ObjectId) { }
    }
}
